### ISSUE TITLE
Koreader is slow on Kpw2

### ISSUE THREAD

**Issue by Steven630** [koreader/koreader#12514]
* KOReader version: 2024.07
* Device: Kindle Paperwhite 2 (firmware: 5.12.2.2)

#### Issue

It could be that the Kindle Paperwhite 2 (KP2) is simply an older device, or it might have something to do with the book or dictionaries. I am reading an epub book with Wordwise and X-ray generated by [WordDumb](https://github.com/xxyzz/WordDumb) and uses pop-up for footnote links. I use the dictionaries quite a lot (#12505 ). When I begin reading, the RAM usage is about 20+ Mb, a few pages on (clicking on footnotes and looking up some words), and within 30 minutes or so, the RAM usage steadily climbs to 100+ Mb and Koreader gets very slow. Sometimes it just freezes when I click on a link or look up a word. I have simplified the CSS files of my dictionaries and it seems to help. Now, instead of getting freezes after the RAM reaches 100 mb, I experience a never-ending "searching dictionary for..." when I use dictionary, and if I click on the blank area, I can cancel the look-up without significant slow-down in page-turning.

I have disabled font kerning and set image scaling to fast. Is there anything else I can do to improve performance? I have also tried to start Koreader with no framework, and it does not help much.
Would installing a version of Koreader for older Kindle devices (I'm using the PW2) make a difference?

#### Steps to reproduce
[sa.zip](https://github.com/user-attachments/files/16995534/sa.zip)
This is the book and its sdr folder. Read it on Kpw2 and click on some footnotes (which appears as pop-ups) and look up words.

##### `crash.log` (if applicable)
[crash.log](https://github.com/user-attachments/files/16995551/crash.log)

I just turned on verbose log today and high RAM usage occurred in the afternoon although this time it did not freeze.

**Comment by NiLuJe** [comment]
> Would installing a version of Koreader for older Kindle devices (I'm using the PW2) make a difference?

No.

----

That's a perfectly reasonably memory consumption for what you're doing. Memory pressure is just *extremely* high on Kindles, especially on devices with low RAM capacity to begin with (I don't recall the PW2 specs exactly, but it might be 256MB).

**Comment by NiLuJe** [comment]
> I have disabled font kerning and set image scaling to fast.

That, on the other hand, is unlikely to make much difference, especially for how badly it affects the user experience.

**Comment by Steven630** [comment]
> > Would installing a version of Koreader for older Kindle devices (I'm using the PW2) make a difference?
> 
> No.
> 
> That's a perfectly reasonable memory consumption for what you're doing. Memory pressure is just _extremely_ high on Kindles, especially on devices with low RAM capacity to begin with (I don't recall the PW2 specs exactly, but it might be 256MB).
> 
> Short of switching platforms & devices, for that workflow, the best you can do is "no framework" mode.

So there's nothing I can do? Do newer Kindle versions behave better? A dictionary look-up for Korean words again triggered a freeze.
[crash (1).log](https://github.com/user-attachments/files/17000033/crash.1.log)

Would a feature described in #12505 help?

**Comment by Commodore64user** [comment]
Top Menu > Cog > Device > Device status alerts > High memory usage

You could force a restart every time RAM usage surpasses a pre-defined threshold level. On my Kindle Keyboard I have it set to 70MB but then again I rarely hit that mark.

**Comment by Steven630** [comment]
> Top Menu > Cog > Device > Device status alerts > High memory usage
> 
> You could force a restart every time RAM usage surpasses a pre-defined threshold level. On my Kindle Keyboard I have it set to 70MB but then again I rarely hit that mark.

Thank you. I have already done that. But since RAM usage also appears in the status bar and the threshold is reached pretty fast,  it's just easier to do a manual reboot.   Your RAM usage doesn't exceed 70 mb? Do you use the dictionary?

**Comment by Steven630** [comment]
If a footnote contains characters of non-English languages (Thai name of a historical Thai kingdom quoted from a Wikipedia entry,for example), the RAM usage increases noticeably. If it's just English words, it seems OK. Do pop-up footnotes store font files?

**Comment by poire-z** [comment]
RAM usage may jump the first time you do a dict lookup with result from HTML dict, or when displaying the first popup footnote - these use MuPDF, and loading MuPDF will occupy some RAM. On my Kobo, after a fresh startup, I jump from 20 Mb to 70 Mb when doing that.
As for displaying text in thai, I have no real idea about how MuPDF goes at finding a font able to display them. It doesn't look like we feed it with our list of fallback fonts, so it may have its own fonts. @benoit-pierre : do you know ?
(And no, popup footnotes don't "store font files", they are just getting HTML text as input, feed it to MuPDF, and get some bitmap to display.)

**Comment by Steven630** [comment]
I am using default fonts and Bookerly for English. Footnotes from Wikipedia are created with  WordDumb, a Calibre plugin. As I am reading a book in English about Southeast Asia, there are footnotes for past kingdoms. So kingdom names in native languages appear in the footnotes ( just kingdom names that appear in brackets for once). Sometimes they are not displayed properly, with squares instead of characters. But even on such occasions, RAM usage jumps. I think this is because of those  non-English characters.  The foodnotes are about the same length,  otherwise there will be no reason for the RAM usage jump.

I also notice that looking up Korean words in dictionary uses more RAM.

**Comment by Commodore64user** [comment]
> Thank you. I have already done that. But since RAM usage also appears in the status bar and the threshold is reached pretty fast,  it's just easier to do a manual reboot. Your RAM usage doesn't exceed 70 mb? Do you use the dictionary? 

Let me rephrase that, that option measures RAM usage for a continuous amount of time (5 mins by default), so if you get to 80MB for one minute it won't notify you. Only after 5 mins of continuous RAM usage above the threshold will it notify you. I suppose that is why i don't get the warnings very often.

**Comment by NiLuJe** [comment]
Define "jump". (And what you're actually checking to begin with, and what kind of values you're seeing).

**Comment by Steven630** [comment]
> Define "jump". (And what you're actually checking to begin with, and what kind of values you're seeing).

RAM usage increases from about 20m to about 70m(on the status bar) if I open pop-up footnotes with non-English characters but stays stable with English footnotes. 

Here are some screenshots:

begins at 26:
![IMG_20240929_011834.jpg](https://github.com/user-attachments/assets/a5b31b02-c4ad-410f-bfde-d31d67e1978a)

footnote for "China" is long, but without specific characters:
![IMG_20240929_012043.jpg](https://github.com/user-attachments/assets/def4a379-2510-4d8d-92c6-c1065479a9b3)

RAM usage 28 after exiting the China footnote:

![IMG_20240929_012244.jpg](https://github.com/user-attachments/assets/541a6114-5bb7-4291-8bc1-1073e0e2d268)

A few pages on, clicking on other all-English footnotes, RAM still at 26. Footnotes for Angkor Wat and Ne Win are shorter than China's but have special characters:

![IMG_20240929_012550.jpg](https://github.com/user-attachments/assets/095a2465-a546-4893-938c-ea12e688284e)

![IMG_20240929_013152.jpg](https://github.com/user-attachments/assets/d52fad93-98fb-4ad3-8bc5-a5ee5dec09d5)


RAM usage increases to 81 after exiting the Angkor Wat and "Ne Win" footnote (both have special characters, ie Burmese):
![IMG_20240929_012724.jpg](https://github.com/user-attachments/assets/e8ec4d8d-a47f-4241-ad95-6b5c0afc39e8)

**Comment by benoit-pierre** [comment]
> RAM usage may jump the first time you do a dict lookup with result from HTML dict, or when displaying the first popup footnote - these use MuPDF, and loading MuPDF will occupy some RAM.

The MuPDF library is loaded unconditionally at startup (by the HTML widget code).

> As for displaying text in thai, I have no real idea about how MuPDF goes at finding a font able to display them. It doesn't look like we feed it with our list of fallback fonts, so it may have its own fonts. @benoit-pierre : do you know ? 

We patch MuPDF to install our own fallback fonts loader:
```c
fz_install_load_system_font_funcs(ctx, NULL, NULL, &fz_load_system_fallback_font_from_file);

[…]

fz_font *
fz_load_system_fallback_font_from_file(fz_context *ctx, int script, int language, int serif, int bold, int italic)
{
	char *filename;
	fz_font *font;

	filename = get_font_file("noto/NotoSansCJKsc-Regular.otf");
	font = fz_new_font_from_file(ctx, NULL, filename, 0, 1);
	free(filename);
	return font;
}
```
Which does ignore the script, this might be improved?

**Comment by poire-z** [comment]
> The MuPDF library is loaded unconditionally at startup (by the HTML widget code).

There is still a big bump: from 27 MB to 63MB just now on my Kobo when I lookup a word and my first dict result is from a HTML dict. When the first result is from a non-HTML dict, it bumps from 27MB to 29MB.

> We patch MuPDF to install our own fallback fonts loader:

Fallback font~~s~~ (singular), it looks like it only has one fallback font, NotoSansCJK ?
Which I guess is the best one for most of the other users in the east :)
FreeSans or FreeSerif (better) would have most of the other scripts coverage. Oh, but it looks like we mention FreeSerif in external_fonts.patch, so we may get Hebrew and some other scripts.

> Which does ignore the script, this might be improved?

I don't think we anywhere associate fonts with scripts, or that we are able to guess what script some text is, or that we would be interested investigating that. At least, I wouldn't be :)

**Comment by benoit-pierre** [comment]
> > Which does ignore the script, this might be improved?
> 
> I don't think we anywhere associate fonts with scripts, or that we are able to guess what script some text is, or that we would be interested investigating that. At least, I wouldn't be :)

There's some detection based on the Unicode range:
```c
int
fz_encode_character_with_fallback(fz_context *ctx, fz_font *user_font, int unicode, int script, int language, fz_font **out_font)
{
	fz_font *font;
	int is_serif = user_font->flags.is_serif;
	int is_italic = user_font->flags.is_italic | user_font->flags.fake_italic;
	int is_bold = user_font->flags.is_bold | user_font->flags.fake_bold;
	int gid;

	gid = fz_encode_character(ctx, user_font, unicode);
	if (gid > 0)
		return *out_font = user_font, gid;

	if (script == 0)
		script = ucdn_get_script(unicode);

	/* Fix for ideographic/halfwidth/fullwidth punctuation forms. */
	if ((unicode >= 0x3000 && unicode <= 0x303F) || (unicode >= 0xFF00 && unicode <= 0xFFEF))
	{
		if (script != UCDN_SCRIPT_HANGUL &&
				script != UCDN_SCRIPT_HIRAGANA &&
				script != UCDN_SCRIPT_KATAKANA &&
				script != UCDN_SCRIPT_BOPOMOFO)
			script = UCDN_SCRIPT_HAN;
	}

	font = fz_load_fallback_font(ctx, script, language, is_serif, is_bold, is_italic);
	if (font)
	{
		gid = fz_encode_character(ctx, font, unicode);
		if (gid > 0)
			return *out_font = font, gid;
	}

[…]
```

**Comment by Steven630** [comment]
But for this hike in RAM usage, I can use my kpw2 for another few years. Just now, a single look-up of "Vientiane", the capital of Laos, with Wiktionary en-en offered by KOreader raised RAM usage from 20+ to 80+.
